modules/concurrency/src/spinner.cpp

```diff
@@ -0,0 +1,55 @@
+//=================================================================================================
+// Copyright (C) 2023-2024 HEPHAESTUS Contributors
+//=================================================================================================
+
+#include "hephaestus/concurrency/spinner.h"
+
+#include <absl/log/log.h>
+#include <fmt/format.h>
+
+#include "hephaestus/base/exception.h"
+
+namespace heph::concurrency {
+Spinner::Spinner() : is_started_(false), stop_requested_(false) {
+}
+
+Spinner::~Spinner() {
+  if (is_started_.load() || spinner_thread_.joinable()) {
+    LOG(FATAL) << "Spinner is still running. Call stop() before destroying the object.";
+    std::terminate();
+  }
+}
+
+void Spinner::start() {
+  throwExceptionIf<InvalidOperationException>(is_started_.load(), fmt::format("Spinner is already started."));
+
+  // NOTE: Replace with std::stop_token and std::jthread when clang supports it.
+  spinner_thread_ = std::thread([this]() { spin(); });
+
+  is_started_.store(true);
+}
+
+void Spinner::spin() {
+  while (!stop_requested_.load()) {
+    spinOnce();
+  }
+}
+
+void Spinner::stop() {
```

Comment:
given that this can take time to terminate maybe we want to return a `std::future`?

Reply:
do we need this to be async? can be blocking, no?

Reply:
it could be nice if we control multiple spinner we can call stop in parallel rather than one by one and waiting each one to terminate

Reply:
done